### PR TITLE
[MRG] MAINT: refactor distal/proximal stuff

### PR DESCRIPTION
**Pull request by jasmainak**
Just a tiny adjustment to the code inspired by #129 . 

We need to develop more high level interfaces but this is a start. The challenge in this PR would be to keep it under control and not touch everything :)

will add a similar method to `L2Pyr` and then basket cells. Then it's good to go

### CODE REVIEW

**Comment by jasmainak** [comment]
@blakecaldwell @rythorpe one round of review here would be appreciated before I go too far down this path. 

The idea is the following: have a high-level method `_connect_at_loc` that lets us connect at proximal or distal dendrites because we always connect in groups of dendrites. This method internally calls `parconnect_from_src` which in turn calls `gid_connect`. 

Now the hanging / loose thread is `_connect`. It also calls `parconnect_from_src` but it's more for cell-to-cell connections. It should probably go away ... I can do it in this PR or another PR, don't want to overwhelm the review process.

File: hnn_core/pyramidal.py [hunk 1]
```diff
@@ -1046,25 +975,18 @@ def parreceive_ext(self, type, gid, gid_dict, pos_dict, p_ext):
...
-                self.ncfrom_extpois.append(
-                    self.parconnect_from_src(
-                        gid_extpois, nc_dict, self.apicaloblique_ampa))
+                self._connect_at_loc(
+                    loc='proximal', receptor='ampa',
+                    gid_src=gid_extpois,
+                    nc_dict=nc_dict,
+                    nc_list=self.ncfrom_extpois)
 
                 if p_ext[self.celltype][1] > 0.0:
```
Comment (jasmainak) [inline hnn_core/pyramidal.py:984]: this if condition is awkward but I didn't want to break things, so left it there

File: hnn_core/pyramidal.py [hunk 2]
```diff
@@ -219,6 +219,43 @@ def _synapse_create(self, p_syn):
...
+
+        postsyns = list()
+        for dend in dends:
+            postsyns.append(getattr(self, f'{dend}_{receptor}'))
+
+        for postsyn in postsyns:
+            nc = self.parconnect_from_src(gid_src, nc_dict, postsyn)
+            nc_list.append(nc)
+
+        return postsyns
```
Comment (rythorpe) [inline hnn_core/pyramidal.py:256]: So the main point here is to simplify the `parreceive()` and `parreceive_ext()` methods correct? If so, I definitely like how this PR focuses on consolidating redundant code.

File: hnn_core/pyramidal.py (hunk 1 above)
Comment (rythorpe) [reply hnn_core/pyramidal.py:983]: Agreed. We'll need to fix this when we remove the loops that iterate through the different feed types from the cell classes. I'm not sure why this was put here in the first place (maybe network build time?) since an nmda receptor conductance of of zero shouldn't have any effect on the simulation.

Comment (jasmainak) [reply hnn_core/pyramidal.py:983]: isn't this related to your PRs about turning off inputs based on the weights? or was that different?

> maybe network build time?

I hope not. "Premature optimization is the root of all evil." (Knuth) Another related aphorism is [make it work, make it nice, make it fast](https://wiki.c2.com/?MakeItWorkMakeItRightMakeItFast)

Network build times are really small compared to simulation time and developer time (most important in my opinion). Maybe for GUI application, some of these times matter but I'd rather refactor first, then profile and worry about timing.

File: hnn_core/pyramidal.py (hunk 2 above)
Comment (jasmainak) [reply hnn_core/pyramidal.py:256]: yes, but also to make the code a bit more readable. Not just concise but more readable. We'll probably need to do a couple more iterations / PRs to get it absolutely on point.

File: hnn_core/pyramidal.py (hunk 1 above)
Comment (rythorpe) [reply hnn_core/pyramidal.py:983]: > isn't this related to your PRs about turning off inputs based on the weights? or was that different?

It's related in principle, but I was unaware that there was a check for synaptic weight >0 within a cell class method. There is probably more to the story that we are unaware of, but we'll need to handle this more consistently in the future.

File: hnn_core/pyramidal.py [hunk 3]
```diff
@@ -219,6 +219,43 @@ def _synapse_create(self, p_syn):
                 self.dends['apical_tuft'](0.5), **p_syn['gabaa'])
 
 
+    def _connect_at_loc(self, loc, receptor, gid_src, nc_dict, nc_list):
```
[the remark this answers is not in the document]
Comment (jasmainak) [reply hnn_core/pyramidal.py:222]: Let me get back to you on that in a day or two. I need to mess around with the code to figure out what's possible and what's not. I think we can definitely extend to `Basket` cells. But I'd leave the refactoring with `_connect` method to another PR. 

So something like `_connect_feed_at_loc` might do ...

[the remark this answers is not in the document]
Comment (rythorpe) [reply hnn_core/pyramidal.py:222]: `_connect_feed_at_loc` sounds good to me! We can always rename it in a later PR while refactoring the cell classes.

**Comment by jasmainak** [comment]
okay I think I'm done here. Couldn't see much scope for improvement in the `Basket` class except this commit 6cd00a8. I can't believe that the `if-else` clause had duplicate statement. I visually verified thrice, but please check :-)

File: hnn_core/pyramidal.py [hunk 4]
```diff
@@ -218,6 +218,43 @@ def _synapse_create(self, p_syn):
...
+        if loc == 'proximal':
+            dends = ['apicaloblique', 'basal2', 'basal3']
+        elif loc == 'distal':
+            dends = ['apicaltuft']
+        else:
+            raise ValueError('loc must be one of proximal or distal')
+
+        postsyns = list()
+        for dend in dends:
+            postsyns.append(getattr(self, f'{dend}_{receptor}'))
```
Comment (rythorpe) [inline hnn_core/pyramidal.py:250]:
```suggestion
        [self.synapse[syn_key] for syn_key in self.synapses.keys() if receptor in syn_key]
```

File: hnn_core/basket.py [hunk 5]
```diff
@@ -149,20 +149,14 @@ def parreceive_ext(self, type, gid, gid_dict, pos_dict, p_ext):
 
                 # connections depend on location of input - why only
                 # for L2 basket and not L5 basket?
-                if p_ext['loc'] == 'proximal':
+                if p_ext['loc'] in ['proximal', 'distal']:
```
[the remark this answers is not in the document]
Comment (jasmainak) [reply hnn_core/basket.py:152]: This is fair, in fact I was looking for something that already existed but couldn't find it. Are we meeting tomorrow afternoon? Should we try to implement this exact thing with pair/triplet programming? I want to keep the discussion more concrete and less "meta". 

I also tagged you in a follow up PR I planned. Should I just append that PR on to this one (it's a fairly straightforward change)? It will make it easier for me to make changes on top without worrying about rebase conflicts. Or would you prefer to review it separately?

[the remark this answers is not in the document]
Comment (rythorpe) [reply hnn_core/basket.py:152]: I see, so https://github.com/jasmainak/hnn-core/pull/1 is basically just moving `parreceive_` methods up to their parent classes. I think this is manageable to merge and review in this PR.

What time are we meeting tomorrow? I'm down for another programming session, though I'm also happy to attempt implementing my proposed changes asynchronously if that would be helpful.

[the remark this answers is not in the document]
Comment (jasmainak) [reply hnn_core/basket.py]: okay I copied what was in #https://github.com/jasmainak/hnn-core/pull/1 into this PR then.

**Comment by jasmainak** [comment]
@rythorpe I implemented what you suggested in 4cf4a2b8059981656c51e7195da63c6ec7702c46

i'm a bit bamboozled that the test fails. Do you see anything in the diff that is off?

**Comment by jasmainak** [comment]
okay found the problem. It's a problem with Neuron, it should raise an error if a `float` is passed. Just lost 2 hours tracking down the issue.

Anyhow, now I have monkeypatched it in f5c891121bae9f88

**Comment by rythorpe** [comment]
> okay found the problem. It's a problem with Neuron, it should raise an error if a `float` is passed. Just lost 2 hours tracking down the issue.
> 
> Anyhow, now I have monkeypatched it in [f5c8911](https://github.com/jonescompneurolab/hnn-core/commit/f5c891121bae9f88b141668ad7e1b0348777b3a8)

Sorry I couldn't be of much help here. I briefly looked things over but arrived at the conclusion that some digging was necessary. So the problem was that the `secloc` argument has to be of an instance of `nrn.Segment`?

**Comment by jasmainak** [comment]
yeah so apparently `Exp2Syn` also accepts floating point numbers. I passed `0.5` instead of `soma(0.5)` by mistake. It should have complained but it didn't.

**Comment by jasmainak** [comment]
I'm pretty much done here btw

File: hnn_core/basket.py [hunk 6]
```diff
@@ -36,51 +47,16 @@ def __set_props(self, cell_name, pos):
...
-    def parconnect(self, gid, gid_dict, pos_dict, p):
-        self._connect(gid, gid_dict, pos_dict, p, 'L2_pyramidal', 'L2Pyr',
-                      postsyns=[self.soma_ampa])
-        self._connect(gid, gid_dict, pos_dict, p, 'L2_basket', 'L2Basket',
-                      lamtha=20., postsyns=[self.soma_gabaa])
+        self.synapses['soma_ampa'] = self.syn_create(
+            self.soma(0.5), e=0., tau1=0.5, tau2=5.)
+        self.synapses['soma_gabaa'] = self.syn_create(
+            self.soma(0.5), e=-80, tau1=0.5, tau2=5.)
+        # this is a pretty fast NMDA, no?
```
Comment (blakecaldwell) [inline hnn_core/basket.py:57]: Should this be clarified rather than having a comment buried in the code?

File: hnn_core/cell.py [hunk 7]
```diff
@@ -162,6 +162,39 @@ def move_to_pos(self):
         """Move cell to position."""
         self.translate_to(self.pos[0] * 100, self.pos[2], self.pos[1] * 100)
 
+    def _connect_feed_at_loc(self, feed_loc, receptor, gid_src, nc_dict,
+                             nc_list):
```
Comment (blakecaldwell) [inline hnn_core/cell.py:166]: What about a test for this new function, in `test_cell.py`? Can a connection be made from two `_Cell` objects b/w the somas?

File: hnn_core/basket.py [hunk 8]
```diff
@@ -147,51 +127,37 @@ def parreceive_ext(self, type, gid, gid_dict, pos_dict, p_ext):
...
                 gid_extgauss = gid + gid_dict['extgauss'][0]
 
-                # gid works here because there are as many pos
-                # items in pos_dict['extgauss'] as there are cells
                 nc_dict = {
                     'pos_src': pos_dict['extgauss'][gid],
                     # index 0 is ampa weight
                     'A_weight': p_ext[self.celltype][0],
-                    'A_delay': p_ext[self.celltype][1],  # index 2 is delay
+                    'A_delay': p_ext[self.celltype][2],  # index 2 is delay
```
Comment (blakecaldwell) [inline hnn_core/basket.py:151]: So was this a bug?

File: hnn_core/basket.py (hunk 6 above)
Comment (jasmainak) [reply hnn_core/basket.py:57]: umm ... what would you like me to clarify here? I actually just copy-pasted the old comment

File: hnn_core/basket.py (hunk 8 above)
Comment (jasmainak) [reply hnn_core/basket.py:151]: interesting, I didn't see this myself. I guess it popped out in the diff when I consolidated the methods from the two classes. It's why we sorely need tests, thanks for pushing me on that. I'll try to make a test for that method.

File: hnn_core/basket.py (hunk 6 above)
Comment (blakecaldwell) [reply hnn_core/basket.py:57]: Comments that ask unanswered questions aren't that useful. Who were you asking in the first place? Are you comparing to other NMDA time constants somewhere?

https://github.com/jonescompneurolab/hnn-core/commit/41c8774ccd4ff56bb4338a70a1489fa65c9a2dd7#diff-faaeadb2c140c4e4e34881bb5f31ae87R52

Comment (jasmainak) [reply hnn_core/basket.py:57]: It's not my comment, cruft from old code: https://github.com/jonescompneurolab/hnn-core/commit/41c8774ccd4ff56bb4338a70a1489fa65c9a2dd7#diff-8a21b8909c8d5e3dc73c03688313de0eL351 :) I can remove it if you want to.

Comment (jasmainak) [reply hnn_core/basket.py:57]: Basically my strategy was to trash anything that was redundant with the code or other comments. But this one seemed a bit unique ...

**Comment by jasmainak** [comment]
@blakecaldwell test added b49005b8a9dee7d501d

File: hnn_core/basket.py (hunk 8 above)
Comment (jasmainak) [reply hnn_core/basket.py:151]: @blakecaldwell you might want to update it in HNN as well: https://github.com/jonescompneurolab/hnn/blob/master/L2_basket.py#L187

Comment (blakecaldwell) [reply hnn_core/basket.py:151]: Thanks @jasmainak

File: hnn_core/basket.py (hunk 6 above)
Comment (blakecaldwell) [reply hnn_core/basket.py:57]: Oh, I see. Makes complete sense why it was copied forward. However, with all that effort spent on where the comment came from, I verified that the time constant is correct. These are the values have been used since at least 2009. See top of page 5: 
https://journals.physiology.org/doi/pdf/10.1152/jn.00535.2009

This comment can be removed.

Comment (jasmainak) [reply hnn_core/basket.py]: awesome, thank you so much for digging into this. I removed the comment in the last commit!

**Comment by jasmainak** [comment]
oops sorry I forgot to check that 🙈. If build times are critical (for GUI I guess?), perhaps we could add a test to ensure that build time is always under x seconds?

@rythorpe merge if you are happy! more PRs on the way :)

File: hnn_core/basket.py [hunk 9]
```diff
@@ -89,45 +64,49 @@ def parreceive(self, gid, gid_dict, pos_dict, p_ext):
...
                     'lamtha': p_src['lamtha'],
                     'threshold': p_src['threshold'],
                     'type_src': 'ext'
                 }
 
-                # NMDA synapse
-                self.ncfrom_common.append(self.parconnect_from_src(
-                    gid_src, nc_dict_nmda, self.soma_nmda))
+                self._connect_feed_at_loc(
+                    feed_loc='proximal', receptor='nmda',
```
Comment (rythorpe) [inline hnn_core/basket.py:95]: Why does `feed_loc='proximal'` in all of the `BasketSingle._connect_feed_at_loc()` calls? I feel like I'm missing something really obvious...

Comment (jasmainak) [reply hnn_core/basket.py:95]: This is a very good point. So looking at the [documentation of HNN](https://jonescompneurolab.github.io/hnn-under_the_hood/04_evoked-rhythmic-inputs/04_evoked-rhythmic-inputs), it says this for proximal:

> The left schematic here shows proximal inputs which target basal dendrites of layer 2/3 and layer 5 pyramidal neurons, and somata of layer 2/3 and layer 5 interneurons.

and this for distal:

> The right schematic shows distal inputs which target the distal apical dendrites of layer 5 and layer 2/3 pyramidal neurons and the somata of layer 2/3 interneurons

*So, for layer 2/3 there is no difference between proximal and distal. For layer 5, you have an input drive to somata for proximal but not distal*

Now if you go the file [L5_basket.py](https://github.com/jonescompneurolab/hnn/blob/master/L5_basket.py) (I'm using the HNN repository since that's our "gold standard" implementation) and look for `'loc'` there is no reference to it. If you go to [L2_basket.py](https://github.com/jonescompneurolab/hnn/blob/master/L2_basket.py#L159-L167), there is an if-else clause but both branches have **exactly** the same lines (right??) with a cryptic comment which asks "why only L2 basket and not L5 basket". Am I missing something?

Comment (blakecaldwell) [reply hnn_core/basket.py:95]: I agree with everything you said @jasmainak. It appears that there is no difference between distal and proximal in the code. It looks like the code changed from only having combined excitatory inputs to both ampa/nmda in the following two commits (in which the comment was added). Commit [317d3324](https://github.com/jonescompneurolab/hnn/commit/317d332407071f1dd5baba66c6bf00ef80c08075#diff-adc0f6cc5cc1a0937da7a9202d463707R159-R164) explains why there was initially an if/else clause. Then the branches became the same in commit [3eee0bd3](https://github.com/jonescompneurolab/hnn/commit/3eee0bd35015a5bbd2bd7761a1c3f1fce8cedd12#diff-adc0f6cc5cc1a0937da7a9202d463707R160-R167).

I think the comment is just reflecting what you observed @jasmainak.

Additionally, a similar comment in [L5_basket.py](https://github.com/jonescompneurolab/hnn/commit/3eee0bd35015a5bbd2bd7761a1c3f1fce8cedd12#diff-0c776e7033b7f5dc6378a67be72276f7R147) noted the difference, but the code still the same thing for proximal and distal inputs. The desired effect comes from there not being a 'L5_basket' key in `p_ext` for distal evoked inputs (in [paramrw.py](https://github.com/jonescompneurolab/hnn/blob/b1522b9e517bc110131ecd2281c9025240980737/paramrw.py#L678-L711)).

Comment (rythorpe) [reply hnn_core/basket.py:95]: Yes, we should expect L2/3 basket cells to receive both proximal and distal inputs and L5 basket cells to receive only proximal inputs.

>If you go to L2_basket.py, there is an if-else clause but both branches have exactly the same lines (right??)

Correct, the if/else statement is unnecessary. It looks like this was corrected in [L5_basket.py](https://github.com/jonescompneurolab/hnn/blob/b1522b9e517bc110131ecd2281c9025240980737/L5_basket.py#L146-L172).

>with a cryptic comment which asks "why only L2 basket and not L5 basket". Am I missing something?

This comment is resolved by noting that HNN code *does*  control for proximal vs. distal inputs for both the L2 and L5 basket cell cases. Both [L5_basket](https://github.com/jonescompneurolab/hnn/blob/b1522b9e517bc110131ecd2281c9025240980737/L5_basket.py#L148) and [L2_basket](https://github.com/jonescompneurolab/hnn/blob/b1522b9e517bc110131ecd2281c9025240980737/L2_basket.py#L138) have an 'if' statement that filters feed location-specific connections according to the parameter keys passed via [`p_ext`](https://github.com/jonescompneurolab/hnn/blob/b1522b9e517bc110131ecd2281c9025240980737/paramrw.py#L678-L711). Particularly note that `p_unique['evdist*']` items don't contain a dict key for `L5_basket`.

Comment (rythorpe) [reply hnn_core/basket.py:95]: Sorry @blakecaldwell your previous post just loaded for me.

File: hnn_core/basket.py [hunk 10]
```diff
@@ -22,6 +31,8 @@ def __init__(self, gid, pos, cell_name='Basket'):
         # for height and xz plane for depth. This is opposite for model as a
         # whole, but convention is followed in this function ease use of gui.
         self.shape_soma()
+        self.synapses = dict()
+        self.sect_loc = dict(proximal=['soma'], distal=[])
```
Comment (rythorpe) [inline hnn_core/basket.py:35]: `self.sect_loc` shouldn't be defined at the `BasketSingle` level, but rather separately for `L2Basket` and `L5Basket` respectively.

Comment (rythorpe) [reply hnn_core/basket.py:35]: I think the reason this is working without throwing any test errors is because by forcing all `feed_loc` variables to be proximal, we're forcing `parreceive_ext` to attempt running `self._connect_feed_at_loc()` if the layer-specific basket cell is defined in `p_src`.

File: hnn_core/basket.py [hunk 11]
```diff
@@ -89,45 +64,49 @@ def parreceive(self, gid, gid_dict, pos_dict, p_ext):
...
-                    gid_src, nc_dict_ampa, self.soma_ampa))
+                self._connect_feed_at_loc(
+                    feed_loc='proximal', receptor='ampa',
+                    gid_src=gid_src, nc_dict=nc_dict,
+                    nc_list=self.ncfrom_common)
 
             # Check if NMDA params are defined in p_src
-            if 'L2Basket_nmda' in p_src.keys():
-                nc_dict_nmda = {
+            if f'{self.name}_nmda' in p_src.keys():
```
Comment (rythorpe) [inline hnn_core/basket.py:84]: As in here (reference to https://github.com/jonescompneurolab/hnn-core/pull/146#discussion_r479299196).

File: hnn_core/basket.py (hunk 9 above)
Comment (jasmainak) [reply hnn_core/basket.py:95]: wow! this sounds like a Sherlock Holmes detective story ... thanks for the sleuthing both of you.

Let me see what I can do to improve the readability.

File: hnn_core/basket.py (hunk 10 above)
Comment (jasmainak) [reply hnn_core/basket.py]: Done. Moving forward, we may want to reduce the `if-else` clauses. They make the code really hard to follow and prone to bugs when changing. Thanks for carefully reviewing it.

**Comment by jasmainak** [comment]
I pushed https://github.com/jonescompneurolab/hnn-core/pull/146/commits/1c71219dc5878014a12e76d36abbdec5ab2f5885. Let me know what you guys think. It doesn't change anything functionally -- just helps with the readability. Because the functionality is taken care through the `if-else` clause but let's reserve consolidating that for another PR.

Also, before you merge, do confirm that I got this right:

- common inputs and evoked inputs can be both proximal and distal
- gaussian and poisson inputs are always proximal

It's how the code in the `pyramidal.py` file appears to be, and I passed `feed_loc` in `basket.py` accordingly -- so the two files match. Just helps with clarity ... functionality as we know is taken care of by the `if-else` clause. I can see more scope for code reduction/improvement but reserving that for later.

**Comment by blakecaldwell** [comment]
> If build times are critical (for GUI I guess?), perhaps we could add a test to ensure that build time is always under x seconds?

I like the idea of a performance regression test. It can be easy to change something that is undetectable on a single core, but blows up during parallelization. I'd actually like to do this for network building and simulations.

**Comment by rythorpe** [comment]
> I pushed [1c71219](https://github.com/jonescompneurolab/hnn-core/commit/1c71219dc5878014a12e76d36abbdec5ab2f5885). Let me know what you guys think. It doesn't change anything functionally -- just helps with the readability. Because the functionality is taken care through the `if-else` clause but let's reserve consolidating that for another PR.

This looks really good, thanks @jasmainak.

> Also, before you merge, do confirm that I got this right:
> 
> * common inputs and evoked inputs can be both proximal and distal

Yes.

> * gaussian and poisson inputs are always proximal

This is true per `create_pext()` in both `hnn.paramrw` and hnn_core.params`, however, we may want to rephrase this in future PRs. I'm not convinced the proximal vs. distal feed location was intended to apply to the gaussian or poisson input from a physiological model perspective.
